aggregator/election_mgr.go

```diff
@@ -0,0 +1,518 @@
+		mgr.Unlock()
+		mgr.metrics.followerResign.Inc(1)
+		return nil
+	}
+	mgr.state = electionManagerResigning
+	mgr.resignWg = &sync.WaitGroup{}
+	resignWg := mgr.resignWg
+	resignWg.Add(1)
+	mgr.Unlock()
+
+	err := mgr.leaderService.Resign(mgr.electionKey)
+	if err != nil {
+		resignWg.Done()
+		mgr.Lock()
+		mgr.state = electionManagerOpen
+		mgr.resignWg = nil
+		mgr.Unlock()
+		err := fmt.Errorf("leader service resign failed: %v", err)
+		mgr.metrics.resignErrors.Inc(1)
+		mgr.logError("resign failed", err)
+		return err
+	}
+
+	// Wait for TTL to give the follower instance some time to take over the
+	// leader role before starting a new campaign after resigning.
+	mgr.sleepFn(time.Duration(mgr.electionOpts.TTLSecs()) * time.Second)
+	resignWg.Done()
+
+	// Now wait for the current instance to become a follower.
+	var (
+		ctxDone    = false
+		isFollower = false
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			ctxDone = true
+		case <-mgr.notifyCh:
+			isFollower = mgr.ElectionState() == FollowerState
+		}
+		if ctxDone || isFollower {
+			break
+		}
+	}
+
+	// If the context expires, we cancel the in progress change and re-check
+	// the election state.
+	if !isFollower {
+		mgr.cancelInProgressChange()
+		isFollower = mgr.ElectionState() == FollowerState
+	}
+
+	mgr.Lock()
+	mgr.state = electionManagerOpen
+	mgr.resignWg = nil
+	mgr.Unlock()
+
+	if isFollower {
+		return nil
+	}
+
+	// If the election state is not follower, it means we failed to establish
+	// the follower state with kv.
+	err = fmt.Errorf("instance resigned but is not follower: %v", ctx.Err())
+	mgr.metrics.resignNotFollower.Inc(1)
+	mgr.logError("resign failed", err)
+	return err
+}
+
+func (mgr *electionManager) Close() error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerOpen {
+		return errElectionManagerExpectOpen
+	}
+	close(mgr.doneCh)
+	mgr.state = electionManagerClosed
+	return nil
+}
+
+func (mgr *electionManager) startElectionLoop() {
+	var campaignStatusCh <-chan campaign.Status
+	notDone := func(int) bool {
+		select {
+		case <-mgr.doneCh:
+			return false
+		default:
+			return true
+		}
+	}
+
+	for {
+		if campaignStatusCh == nil {
+			mgr.RLock()
+			state := mgr.state
+			resignWg := mgr.resignWg
+			mgr.RUnlock()
+
+			// NB(xichen): need to wait for the result of the leader service resign call
+			// because resigning is done in a different goroutine and when/whether we need
+			// to restart the campaign depends on the result of the resign call.
+			if state == electionManagerResigning {
+				resignWg.Wait()
+			}
+
+			// NB(xichen): campaign retrier retries forever until either the Campaign call succeeds,
+			// or the election manager has resigned.
+			if err := mgr.campaignRetrier.AttemptWhile(notDone, func() error {
+				var err error
+				campaignStatusCh, err = mgr.leaderService.Campaign(mgr.electionKey, mgr.campaignOpts)
+				if err == nil {
+					return nil
+				}
+				mgr.metrics.campaignCreateErrors.Inc(1)
```

Comment:
Yes this is what `campaignRetrier` is for. I currently use an initial backoff of 100ms and a backoff factor of 2 with a max backoff of 2s for the retrier.